File: Python/Authored/Mastermind.py
import random
def mastermind(game):
    colours = ["Red", "Blue", "Green", "Orange", "Purple", "Yellow"]
    amount = {"Red": 0, "Blue": 0, "Green": 0, "Orange": 0, "Purple": 0, "Yellow": 0, "": 0}
    guess = []
    for colour in colours:
        answer = game.check([colour, colour, colour, colour])
        colour_amount = answer.count("Black")
        amount[colour] = colour_amount
        for a in range(0,colour_amount):
            guess.append(colour)
    answer = False
    while answer != "WON!":
        answer = game.check(random.sample(guess, len(guess)))

import random
class MasterMind():
    def __init__(self):
        self.colours = ["Red", "Blue", "Green", "Orange", "Purple", "Yellow"]
        self.goes = 0
        self.result = []
        for a in range(1,5):
            self.result.append(self.colours[random.randint(0,len(self.colours)-1)])
        self.amount = {"Red": 0, "Blue": 0, "Green": 0, "Orange": 0, "Purple": 0, "Yellow": 0, "": 0}
        for colour in self.result:
            self.amount[colour] += 1

    def check(self, attempt):
        self.goes += 1
        if len(attempt) != 4: return ["Error"]
        if attempt == self.result:
            return "WON!"
        amount = self.amount
        result = []
        new_attempt = []
        for colour in attempt:
            new_attempt.append(colour)
        index = -1
        for colour in attempt:
            index += 1
            if colour not in self.colours:
                return "Error"
            if colour == self.result[index]:
                result.append("Black")
                amount[colour] -= 1
                new_attempt[index] = ""
        for colour in new_attempt:
            if amount[colour] > 0:
                result.append("White")
                amount[colour] -= 1
        return random.sample(result, len(result))

game = MasterMind()
mastermind(game)

File: Ruby/Authored/Mastermind.rb
def mastermind(game)
  colours = ["Red", "Blue", "Green", "Orange", "Purple", "Yellow"]
  amount = Hash.new {|key,value| key[value]=0}
  guess = []
  colours.each do |colour|
    answer = game.check([colour, colour, colour, colour])
    colour_amount = answer.count("Black")
    amount[colour] = colour_amount
    (1..colour_amount).each{|c| guess << colour}
  end
  answer = false

  while answer != "WON!" && answer != "Error"
    p answer = game.check(guess.shuffle)
  end
end

class MasterMind

  Colours = ["Red", "Blue", "Green", "Orange", "Purple", "Yellow"]

  def initialize()
    @result = []
    @goes = 0
    (1..4).each do |colour|
      @result << Colours[rand(Colours.length)]
    end
    @amount = Hash.new {|key,value| key[value]=0}
    @result.each{|colour| @amount[colour]+=1}
  end

  def check(attempt)
    return "Error" if attempt.length != 4
    @goes += 1
    p @goes
    return "Error" if @goes>60
    return "WON!" if attempt == @result
    amount = {"Red"=>0, "Blue"=>0, "Green"=>0, "Orange"=>0, "Purple"=>0, "Yellow"=>0}
    @amount.each{|key,value| amount[key]=value}
    result = []
    new_attempt = []
    attempt.each {|colour| new_attempt << colour}
    attempt.each_with_index do |colour, index|
      return "Error" if !Colours.include? colour
      if colour == @result[index]
        result << "Black"
        amount[colour] -= 1
        new_attempt[index] = ""
      end
    end
    new_attempt.each do |colour|
      if colour != ""
        if amount[colour] > 0
          result << "White"
          amount[colour] -= 1
        end
      end
    end
    result.shuffle
  end
end

game = MasterMind.new
mastermind(game)
